cameo/strain_design/heuristic/evolutionary/optimization.py

```diff
@@ -806,3 +845,223 @@ def __init__(self, reaction_db=None, *args, **kwargs):
+      2013): 236–46. ---doi:10.1089/ind.2013.0005.
+
+    Parameters
+    ----------
+    model : SolverBasedModel
+       the model to operator on
+    cofactor_id_swaps : list
+       a list of length 2 where the first element contains a list of metabolite identifiers to swap-out by replacing
+       them with the metabolites listed in the second element of the same list, see e.g. `NADH_NADPH` which is also
+       the default
+    candidate_reactions : list
+       reactions to consider for co-factor swap - if not given then search for all reactions that include the given
+       cofactors
+    skip_reactions : list
+       reactions to not consider for co-factor swap, defaults to the objective function if not provided
+    args, kwargs : keyword arguments
+       passed on to super-classes, see in particular `objective_function`, `heuristic_method`, `termination`,
+       `simulation_method`, `wt_reference`, of `HeuristicOptimization` and `max_size` of `HeuristicOptimization.run`
+
+    Examples
+    --------
+    >>> from cameo import models
+    >>> model = models.bigg.iJO1366
+    >>> from cameo.strain_design.heuristic.evolutionary.objective_functions import biomass_product_coupled_yield
+    >>> py = biomass_product_coupled_yield(model.reactions.EX_cys__L_e, model.reactions.EX_glc__D_e)
+    >>> swap_optimization = CofactorSwapOptimization(model=model, objective_function=py)
+    >>> swap_optimization.run(max_evaluations=2000, max_size=3)
+    """
+
+    def __init__(self, model, cofactor_id_swaps=NADH_NADPH, candidate_reactions=None,
+                 skip_reactions=None, *args, **kwargs):
+        super(self.__class__, self).__init__(model=model, knockout_evaluator=SwapEvaluator, *args, **kwargs)
+        self.model = SwapperModel(model, cofactor_id_swaps, candidate_reactions, skip_reactions)
+        self._ko_type = SWAP_KNOCKOUT_TYPE
+        self.representation = list(self.model.swapped_reactions.keys())
+        self._decoder = decoders.ReactionKnockoutDecoder(self.representation, self.model)
+
+
+class SwapperModel(SolverBasedModel):
+    """
+    An extension of the `SolverBasedModel` to add functions to easily add reactions that have co-factors swapped etc.
+
+    Attributes
+    ----------
+    model : SolverBasedModel
+       the model to convert
+    cofactor_id_swaps : list
+       a list of length 2 where the first element contains a list of metabolite identifiers to swap-out by replacing
+       them with the metabolites listed in the second element of the same list, see e.g. `NADH_NADPH` which is also
+       the default
+    candidate_reactions : list
+       reactions to consider for co-factor swap - if not given then search for all reactions that include the given
+       cofactors
+    skip_reactions : list
+       reactions to not consider for co-factor swap, defaults to the objective function if not provided
+    """
+    def __init__(self, model, cofactor_id_swaps, candidate_reactions=None, skip_reactions=None):
```

Comment:
Shouldn't we automatically figure out the candidate reactions automatically based on the swap pattern?